Add all functional

PatreevIgor opened the pull request.

serg-kovalev commented on .gitignore:
```diff
@@ -0,0 +1 @@
+*.env
```
why did you put `*` before `.env`?

PatreevIgor replied:
This means that all files with this extension will be in gitignore. In my case * not needed.

serg-kovalev commented on .ruby-version:
```diff
@@ -0,0 +1 @@
+ruby-2.2.2
```
version format is `2.2.2`, without `ruby-`
by the way, why not use 2.3.x ruby? It has more great features

PatreevIgor replied:
Early i deployed applications to locum (russian hosting). Locum used latest version is 2.2.2. On the future i will use latest version ruby.

serg-kovalev commented on README.md:
```diff
@@ -1,2 +1,3 @@
 # sinatra_app
 Sample sinatra application
+Heroku link: https://supersinatra.herokuapp.com/
```
it would be good also to provide information on how to setup application locally (not required at the moment, just suggestion for future projects)
Imagine, that I wanna get started and setup it locally, I shouldn't deep dive into source code to understand what is required to run it in development/production mode.

serg-kovalev replied:
please add description, what your application do
I didn't get it.

PatreevIgor replied:
Ok. In future i will be give comments how to install and use application on the local pc. Now I will add description to my Project.

serg-kovalev commented on Rakefile:
```diff
@@ -0,0 +1,3 @@
+require 'sinatra/activerecord'
+require 'sinatra/activerecord/rake'
+require './app'
```
new line

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,81 @@
+  raise 'You must specify the G_API_SECRET env veriable' unless ENV['OAUTH2_CLIENT_SECRET']
+end
+
+check_g_api_client
```
looks weird. Try to rework it with initializers

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,81 @@
+  begin
+  User.last.delete
+  rescue Exception => e
+    puts e.message
```
that's good for debugging and troubleshooting
as end user, what will I see?

Did you want to put it in logs?

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,81 @@
+
+post '/delete' do
+  begin
+  User.last.delete
```
anyone, even unauthorized/anonymous user can delete last user
That's bad design. Please refactor

PatreevIgor replied:
Ok. I will make it.

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,81 @@
+  uri = URI.parse(request.url)
+  uri.path = '/oauth2callback'
+  uri.query = nil
+  uri.to_s
```
all 4 lines above looks weird.
Please refactor it in one-liner with string interpolation, something like this:
```ruby
URI("#{APP_BASE_URL}/oauth2callback")
```
name of this method is confusing a bit. I had to search in file where did you use it.
I would remove this method.
You can have just two constants: `APP_BASE_URL` and `GOOGLE_OAUTH_CALLBACK_URL` at the top of application.
I would recommend to have a separate class that will handle authorization/authentication things.

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,81 @@
+                :authorize_url => "/o/oauth2/auth",
+                :token_url => "/o/oauth2/token"
+              })
+end
```
move it to class as described above

serg-kovalev commented on views/_status_user.erb:
```diff
@@ -0,0 +1 @@
+<%= session[:access_token] ? session[:email]['data']['email'] : 'You are nor autorized' %>
```
I would call a helper with a name `authorized?` or something like that

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,81 @@
+set :partial_template_engine, :erb
+enable :sessions
+
+SCOPES = ['https://www.googleapis.com/auth/userinfo.email'].join(' ')
```
strange name of constant

serg-kovalev commented on app.rsa:
```diff
@@ -0,0 +1,27 @@
+aGc+XQKBgGRp46/J3Y+7YX9o9ESjAxmzLNzWG2r/PqLNsT30ri/1lfzzNEWqjmvb
+jAk0moHPUAjlxLT63LHXqqce0vA2vXv6ufmCKmT2TerPArjPPWimPeslKqEQfyCx
+m+/Ky2I1JPNSjagrgLPi+9vYu+Go0TWMabxeYJ+vrcB2Kf/Dqg6A
+-----END RSA PRIVATE KEY-----
```
public & private keys in repo - why?
it's not secure

serg-kovalev replied:
It is still here

serg-kovalev commented on app.rsa.pub:
```diff
@@ -0,0 +1,9 @@
+7y7E2b1hgW0AUbWfRB8dJZtc1tdYxLDAVvuzU31ARnpMXhYYEXUXrD2sSubBhppW
+wge38zb5ODa7Du/X2GhoN5miGdyOG2KS8nTLu+zvg7I9VNJAkJijqLtbF5TFKk7r
+XQIDAQAB
+-----END PUBLIC KEY-----
```
public & private keys in repo - why?
it's not secure

serg-kovalev replied:
It is still here

serg-kovalev commented on config/database.yml:
```diff
@@ -0,0 +1,13 @@
+development:
+  database: sinatra_development
+  adapter: postgresql
+  user: patron
```
use ENV variables for user, host, password and port

serg-kovalev commented on views/layout.erb:
```diff
@@ -0,0 +1,45 @@
+<!DOCTYPE html>
+<html>
+	<head>
```
styly guide
replace tabs with spaces, pay attention to indent

PatreevIgor commented:
I'm not ignoring your comments. I was instructed to do in the first place team project. I just today managed to finish everything.
Before perform the task I agreed with mentors that I will do authentication via Google, because I didn't do VK and GH autentification.
I fixed the following comments:
- I added JWT
- I fixed the comments on the design
- I added Rubocop
- I changed bad names of the commit
- I added description to my Project (to Readme.md)
- I created new class 
- I transferred secret information to git ignore
- I added flash message to exeption...
I apologize that I did for so long and I hope I haven't forgotten to fix.
Look now please.

PatreevIgor commented:
app.rsa.pub, app.rsa and database.yml deleted from github.

serg-kovalev commented on README.md:
```diff
@@ -1,2 +1 @@
-# sinatra_app
-Sample sinatra application
+The project allows you to authenticate via Google. Registered users can work with the database and can see the JWT before encoding and after. Also registered users can see the contents of the session on the main page.
```
where is the link to application?
Have you deployed it after applying the changes?

serg-kovalev commented on authentificator.rb:
```diff
@@ -0,0 +1,34 @@
+  SITE          = 'https://accounts.google.com'.freeze
+  AUTHORIZE_URL = '/o/oauth2/auth'.freeze
+  TOKEN_URL     = '/o/oauth2/token'.freeze
+  APP_BASE_URL  = 'http://localhost:4567'.freeze
```
you're using localhost in callback url
I'm sure it will not work when you deploy it to heroku. Right?
you will need to load APP_BASE_URL from ENV variable and do setup config at heroku with a proper value for that var.

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,94 @@
+post '/delete' do
+  begin
+    User.last.delete
+  rescue Exception => e
```
rescue form Exception is bad practice. In case of syntax error - you will never see it

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,94 @@
+
+post '/delete' do
+  begin
+    User.last.delete
```
strange behavior with deleting the last user.
Okay, let it be so.
But you don't check if user is authorized. That's not secure.

serg-kovalev commented on app.rb:
```diff
@@ -0,0 +1,94 @@
+end
+
+post '/create' do
+  User.create(first_name: params[:username])
```
just in case there would be an error.
Why not use ActiveRecord::Error class to gracefully show them?

PatreevIgor commented:
I was not able to deploy the application to heroku. On local machine everything works fine. In the logs on heroku there is nothing, I can't understand why.
